pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/MultiPartitionCommitter.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.spark;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.internal.wire.Committer;
+import com.google.cloud.pubsublite.internal.wire.CommitterBuilder;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import com.google.common.util.concurrent.MoreExecutors;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class MultiPartitionCommitter implements Serializable {
+  private static final GoogleLogger log = GoogleLogger.forEnclosingClass();
+
+  public interface CommitterFactory extends Serializable {
+    Committer newCommitter(PslDataSourceOptions options, Partition partition);
+  }
+
+  private final PslDataSourceOptions options;
+  private final CommitterFactory committerFactory;
+  private final ReentrantLock lock = new ReentrantLock();
+  private boolean initialized = false;
+  private final Map<Partition, Committer> committerMap = new HashMap<>(); // lazily constructed
+
+  public MultiPartitionCommitter(PslDataSourceOptions options) {
+    this(
+        options,
+        (opt, partition) ->
+            CommitterBuilder.newBuilder()
+                .setSubscriptionPath(opt.subscriptionPath())
+                .setPartition(partition)
+                .setServiceClient(opt.newCursorClient())
+                .build());
+  }
+
+  @VisibleForTesting
+  public MultiPartitionCommitter(PslDataSourceOptions options, CommitterFactory committerFactory) {
+    this.options = options;
+    this.committerFactory = committerFactory;
+  }
+
+  private void init(Set<Partition> partitions) {
```

Comment:
Yes this won't handle repartition, but we don't have repartition yet. I would start with the assumption that it's immutable for now, and update this later if needed.

Reply:
Right, but... then why is there an "init" method? not just a constructor parameter? There's no need to do two phase construction.

Reply:
I don't have the actual number of partitions when this is constructed. Otherwise I will need to do subscription call to PSL when in PslContinuousReader construction. I would prefer lazily set up once we have that information.

Reply:
Just make this call before continuous reader construction and pass it in?

Reply:
Done.